### PR TITLE
[Improvement] Support for candidate multiple dimensional search constraints.

### PR DESCRIPTION
**Pull request by aptsunny**
Part of [Autoformer algo](https://github.com/open-mmlab/mmrazor/pull/288)

## Motivation
Continued improvements of EvolutionSearchLoop throught multi-resources constraints.

## Use cases (Optional)

Modify the search configs file as follow.

```python
flops_range=(0., 330.) ->  constraints_range=dict(flops=(0., 330.)),
```

```python
train_cfg = dict(
    _delete_=True,
    type='mmrazor.EvolutionSearchLoop',
    dataloader=_base_.val_dataloader,
    evaluator=_base_.val_evaluator,
    max_epochs=20,
    num_candidates=20,
    top_k=10,
    num_mutation=5,
    num_crossover=5,
    mutate_prob=0.2,
    constraints_range=dict(params=(0, 55), flops=(0., 330.)),
    dump_derived_mutable=True,
    score_key='accuracy/top1')
```
## Checklist

**After PR**:

- [ ] The search configuration file for the search algorithm needs to be changed. Please fix it

### CODE REVIEW

File: mmrazor/engine/runner/evolution_search_loop.py [hunk 1]
```diff
@@ -99,7 +105,7 @@ def __init__(self,
                 correct init candidates file'
 
         self.top_k_candidates = Candidates()
-        if resource_estimator_cfg is None:
+        if resource_estimator_cfg:
```
Comment (gaoyang07) [inline mmrazor/engine/runner/evolution_search_loop.py:108]: Should it be `if resource_estimator_cfg is None`?

File: mmrazor/engine/runner/evolution_search_loop.py [hunk 2]
```diff
@@ -214,8 +251,19 @@ def gen_crossover_candidates(self) -> List:
 
             crossover_candidate = self._crossover()
 
-            if self._check_constraints(random_subnet=crossover_candidate):
+            flag = False
```
Comment (gaoyang07) [inline mmrazor/engine/runner/evolution_search_loop.py:254]: reify the `flag` here as it suits the particular judgment of model constraints.

File: mmrazor/structures/subnet/candidate.py [hunk 3]
```diff
@@ -49,12 +57,14 @@ def subnets(self) -> List[Dict]:
         return [item[0] for item in self.data]
 
     def _format(self, data: Any) -> _format_return:
-        """Transform [any, ...] to [tuple(any, float), ...] Transform any to
-        tuple(any, float)."""
+        """Transform [any, ...] to [tuple(any, float, float), ...] Transform
+        any to tuple(any, float, float)."""
 
         def _format_item(item: Any):
             """Transform any to tuple(any, float)."""
```
Comment (gaoyang07) [inline mmrazor/structures/subnet/candidate.py:64]: modify docstring here.

File: mmrazor/engine/runner/evolution_search_loop.py (hunk 1 above)
Comment (aptsunny) [reply mmrazor/engine/runner/evolution_search_loop.py:108]: default resource_estimator_cfg is dict.

Comment (gaoyang07) [reply mmrazor/engine/runner/evolution_search_loop.py:108]: Then directly use `self.estimator = ResourceEstimator(**resource_estimator_cfg)`

File: mmrazor/engine/runner/utils/check.py [hunk 4]
```diff
@@ -15,22 +13,23 @@
...
     """
-    if flops_range is None:
-        return True
+    if constraints_range is None:
+        return True, None
 
     assert hasattr(model, 'set_subnet') and hasattr(model, 'architecture')
     model.set_subnet(subnet)
-    fix_mutable = export_fix_subnet(model)
+    fix_mutable = export_fix_subnet(model, dump_derived)
```
Comment (gaoyang07) [inline mmrazor/engine/runner/utils/check.py:32]: remove `dump_derived`

File: mmrazor/engine/runner/evolution_search_loop.py [hunk 5]
```diff
@@ -214,8 +246,19 @@ def gen_crossover_candidates(self) -> List:
...
```
Comment (gaoyang07) [inline mmrazor/engine/runner/evolution_search_loop.py:249]: rename `flag`

File: mmrazor/structures/subnet/candidate.py [hunk 6]
```diff
@@ -1,33 +1,41 @@
...
 from collections import UserList
-from typing import Any, Dict, List, Optional, Tuple, Union
+from typing import Any, Dict, List, Optional, Union
 
 
 class Candidates(UserList):
-    """The data structure of sampled candidate. The format is [(any, float),
-    (any, float), ...].
+    """The data structure of sampled candidate. The format is Union[Dict[Any,
+    Dict], List[Dict[Any, Dict]]].
```
Comment (sunnyxiaohu) [inline mmrazor/structures/subnet/candidate.py:8]: The format is `List[Dict[Any, Dict]]`

File: mmrazor/structures/subnet/candidate.py [hunk 7]
```diff
@@ -41,23 +49,37 @@ def __init__(self, initdata: Optional[Any] = None):
...
     @property
     def subnets(self) -> List[Dict]:
         """The subnets of candidates."""
-        return [item[0] for item in self.data]
+        return [eval(key) for item in self.data for key, _ in item.items()]
 
-    def _format(self, data: Any) -> _format_return:
-        """Transform [any, ...] to [tuple(any, float), ...] Transform any to
-        tuple(any, float)."""
+    def _format(self, data: Any):
```
Comment (sunnyxiaohu) [inline mmrazor/structures/subnet/candidate.py:70]: update type-hint for data with `Union[List[Dict], Dict].`

**Comment by sunnyxiaohu** [comment]
update type-hint for Candidate::append, extend, insert. It seems that they must accept Dict as input.

File: mmrazor/structures/subnet/candidate.py (hunk 6 above)
Comment (aptsunny) [reply mmrazor/structures/subnet/candidate.py:8]: done

File: mmrazor/structures/subnet/candidate.py (hunk 7 above)
Comment (aptsunny) [reply mmrazor/structures/subnet/candidate.py]: > update type-hint for data with `Union[List[Dict], Dict].`

done

**Pull request merged**